### ISSUE TITLE
Figure out a way to get a list of fields for a content type via REST

### ISSUE THREAD

**Issue by mjordan** [mjordan/islandora_workbench#6]
If we can inspect the fields that a content type has configured, we can validate that the columns in the input file match machine field names, and we can also dynamically create requests for string, taxo term, and linked references.

**Comment by mjordan** [comment]
Unfortunately, `curl -v "admin:islandora@localhost:8000/entity/node_type/islandora_object"` doesn't return a list of fields:

`
{"uuid":"c720394c-5f28-4290-be53-88bc7de9c7bc","langcode":"en","status":true,"dependencies":{"enforced":{"module":["islandora_defaults"]},"module":["menu_ui"]},"third_party_settings":{"menu_ui":{"available_menus":["main"],"parent":"main:"}},"name":"Repository Item","type":"islandora_object","description":"An item in your Islandora repository.","help":"","new_revision":true,"preview_mode":1,"display_submitted":true}
`

JSON::API also provides a way to `GET` config entities, and provides this example:

`
curl \
  --header 'Accept: application/vnd.api+json' \
  http://admin:admin@drupal.d8/jsonapi/menu/menu
`

Can't find any documentation though.

**Comment by mjordan** [comment]
JSON::API to the rescue:

`curl  --header 'Accept: application/vnd.api+json' -o test2.json  http://admin:islanodra@localhost:8000/jsonapi/node/islandora_object?page[limit]=1`

will return one node of the islandora_object type, with all fields listed and populated. Doesn't give you the field type, though.

**Comment by mjordan** [comment]
JSON::API allows fetching of field configuration by field UUID:

`
http://localhost:8000/jsonapi/field_config/field_config/8d5c9dc2-1823-43c9-8c6f-5fae6ae77ec4
`

So, if we can get field UUIDs from content types, we can introspect a node's fields prior to generating requests. Only missing piece is to get the field UUIDs on a node.

**Comment by mjordan** [comment]
Looks like `http://admin:islandora@localhost:8000/jsonapi/field_config/field_config` provides all the fields and their configuration info.

**Comment by mjordan** [comment]
`http://admin:islandora@localhost:8000/jsonapi/field_config/field_config` provides all the fields and their configuration info.

```python
#!/usr/bin/env python3

import os
import sys
import json
import requests

fields_url = 'http://admin:islandora@localhost:8000/jsonapi/field_config/field_config'

headers = {'Content-Type': 'application/json'}
response = requests.get(fields_url, auth=('admin', 'islandora'), headers=headers)

field_config = json.loads(response.text)
field_types = {}
for item in field_config['data']:
	field_name = item['attributes']['field_name']
	field_types[field_name] = item['attributes']['field_type']

print(field_types)
```

gets us a dictionary like this:

```python
{
'body': 'text_with_summary',
'comment_body': 'text_long',
'field_access_terms': 'entity_reference',
'field_file_size': 'integer', 
'field_media_audio_file': 'file',
'field_media_of': 'entity_reference',
'field_media_use': 'entity_reference',
'field_mime_type': 'string',
'field_media_file': 'file',
'field_height': 'integer',
'field_media_image': 'image',
'field_width': 'integer',
'field_media_oembed_video': 'string',
'field_media_video_file': 'file',
'comment': 'comment',
'field_image': 'image',
'field_tags': 'entity_reference',
'field_alternative_title': 'string',
'field_description': 'string_long',
'field_display_hints': 'entity_reference',
'field_edtf_date': 'edtf',
'field_edtf_date_created': 'edtf',
'field_edtf_date_issued': 'edtf',
'field_extent': 'string',
'field_identifier': 'string',
'field_linked_agent': 'typed_relation',
'field_member_of': 'entity_reference',
'field_model': 'entity_reference',
'field_pid': 'string',
'field_resource_type': 'entity_reference',
'field_rights': 'string',
'field_subject': 'entity_reference'
}
```

**Comment by mjordan** [comment]
A glitch. This doesn't provide the cardinality of a field. This is important since PATCHing replaces all values in a field. If a multivalued field already has a value, and we want to keep it and add a second value, we need to get the original value and include it in our PATCH. If a field is monovalued, we can only include one value in our PATCH.

**Comment by mjordan** [comment]
`http://admin:islandora@localhost:8000/jsonapi/field_storage_config/field_storage_config` provides cardinality.

**Comment by mjordan** [comment]
Resolved with 3e952225a4c97a2a45b9872ad96ae813c6e147cb.
